#!/bin/env python3
from gwf import AnonymousTarget
import os, glob

########################## Functions ##########################

def species_abbreviation(species_name: str) -> str:
	"""Creates species abbreviation from species name.

	:param str species_name:
		Species name written as *genus* *species*"""
	genus, species = species_name.replace(' ', '_').split('_')
	genus = genus[0].upper() + genus[1:3]
	species = species[0].upper() + species[1:3]
	return genus + species

def parse_fasta(fasta_file: str):
	"""
	Parses :format:`FASTA` file returning all sequence names and lengths paired in a list of dictionaries.
	
	::
	
		return [{'sequence_name': str, 'sequence_length': int}, ...]
	
	:param str fasta_file:
		Sequence file in :format:`FASTA` format.
	"""
	fasta_list = []
	seq_name = None
	length = 0
	with open(fasta_file, 'r') as fasta:
		for entry in fasta:
			entry = entry.strip()
			if entry.startswith(">"):
				if seq_name:
					fasta_list.append({'sequence_name': seq_name, 'sequence_length': length})
					length = 0
				entry = entry.split(" ", 1)
				seq_name = entry[0][1:]
			else:
				length += len(entry)
		fasta_list.append({'sequence_name': seq_name, 'sequence_length': length})
	return fasta_list

def padding_calculator(parse_fasta: list, size: int  | None = 500000):
	"""
	Calculates proper 0 padding for numbers in **partition_chrom**.

	:param list parse_fasta:
		List of dictionaries produced by **parse_fasta**.
	:param int size:
		Size of partitions. Default 500kb.
	"""
	num = 1
	for chrom in parse_fasta:
		whole_chunks = chrom['sequence_length'] // size
		num += (whole_chunks + 1)
	return len(str(num))

def partition_chrom(parse_fasta: list, size: int = 500000, npad: int = 5):
	"""
	Partitions :format:`FASTA` file parsed with **parse_fasta**.
	
	Uses the list of dictionaries from **parse_fasta** to creates a list of dictionaries
	containing partition number, sequence name, start and end postion (0 based).
	By default the partition size is 500kbs.
	
	::
	
		return [{'num': int, 'region': str, 'start': int, 'end': int}, ...]
	
	:param list parse_fasta:
		List of dictionaries produced by **parse_fasta**.
	:param int size:
		Size of partitions. Default 500kb.
	"""
	chrom_partition = []
	num = 1
	for chrom in parse_fasta:
		whole_chunks = chrom['sequence_length'] // size
		partial_chunk = chrom['sequence_length'] - whole_chunks * size
		start = 0
		for chunk in range(whole_chunks):
			end = start + size
			chrom_partition.append({'num': f'{num:0{npad}}', 'region': chrom['sequence_name'], 'start': start, 'end': end})
			start = end
			num += 1
		if partial_chunk:
			chrom_partition.append({'num': f'{num:0{npad}}', 'region': chrom['sequence_name'], 'start': start, 'end': start + partial_chunk})
			num += 1
	return chrom_partition

def mpileup_partitions_filelist(partitions: list, top_dir: str, species_name: str, npad: int = 5):
	"""
	Generates list of dictionaries containing names of contigs and filenames of :format:`mpileup`-parts per contig based on **partition_chrom** output.

	Uses the list of dictionaries generated by **partition_chrom** to create a list of dictionaries where each dictionary contains the name of a contig
	and the names of output files created **mpileup_part** using those same partitions.

	::

		return [{'contig': str, 'mpileup_files': list}, ...]
	
	:param list partitions:
		List of dictionaries produced by **partition_chrom**.
	:param str top_dir:
		Top most level of current working directory.
	:param str species_name:
		Name of species currently being worked on.
	"""
	mpileup_filelists = []
	contig_name = None
	part_num = 0
	for partition in partitions:
		part_num += 1
		if partition['region'] != contig_name:
			if contig_name:
				mpileup_filelists.append({'contig': contig_name, 'mpileup_files': [f'{top_dir}/tmp/mpileup/{species_abbreviation(species_name)}_{num:0{npad}}_{contig_name}.mpileup' for num in range(start, part_num)]})
			start = part_num
			contig_name = partition['region']
	mpileup_filelists.append({'contig': contig_name, 'mpileup_files': [f'{top_dir}/tmp/mpileup/{species_abbreviation(species_name)}_{num:0{npad}}_{contig_name}.mpileup' for num in range(start, part_num + 1)]})
	return mpileup_filelists

########################## Freebayes ##########################

def name_freebayes_chrom(idx: str, target: AnonymousTarget) -> str:
	return f'{os.path.basename(target.outputs["vcf"].replace("-", "_"))}'

# def freebayes_chrom(reference_genome_file: str, bam_file_list: list, output_directory: str, species_name: str, region: str, num: int, start: int, end: int, ploidy: int = 100, best_n_alleles: int = 3, min_alternate_fraction: float | int = 0, min_alternate_count: int = 2):
# 	"""
# 	Template: Create VCF file for each 'chromosome' in pooled alignment.
	
# 	Template I/O::
	
# 		inputs = {}
# 		outputs = {}
	
# 	:param
# 	"""
# 	bam_file_string = ' -b '.join(bam_file_list)
# 	inputs = {'reference': reference_genome_file,
# 		   	  'bam': bam_file_list}
# 	outputs = {'vcf': f'{output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.vcf'}
# 	options = {
# 		'cores': 1,
# 		'memory': '100g',
# 		'walltime': '36:00:00'
# 	}
# 	spec = f"""
# 	# Sources environment
# 	if [ "$USER" == "jepe" ]; then
# 		source /home/"$USER"/.bashrc
# 		source activate vcf
# 	fi
	
# 	export _JAVA_OPTIONS="-Xmx{options['memory']}"

# 	echo "START: $(date)"
# 	echo "JobID: $SLURM_JOBID"

# 	[ -d {output_directory}/freebayes_vcf/tmp ] || mkdir -p {output_directory}/freebayes_vcf/tmp
	
# 	freebayes \
# 		-f {reference_genome_file} \
# 		-n {best_n_alleles} \
# 		-p {ploidy} \
# 		-r {region}:{start}-{end} \
# 		--min-alternate-fraction {min_alternate_fraction} \
# 		--min-alternate-count {min_alternate_count} \
# 		--pooled-discrete \
#         -b {bam_file_string} \
# 	| bcftools filter \
# 		-e 'INFO/TYPE~"del" || INFO/TYPE~"ins"' \
# 		-O v \
# 		- \
# 		> {output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.prog.vcf
	
# 	mv {output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.prog.vcf {outputs['vcf']}
	
# 	echo "END: $(date)"
# 	echo "$(jobinfo "$SLURM_JOBID")"
# 	"""
# 	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def freebayes_chrom(reference_genome_file: str, bam_file_list: list, output_directory: str, species_name: str, region: str, num: int, start: int, end: int, ploidy: int = 100, best_n_alleles: int = 3, min_alternate_fraction: float | int = 0, min_alternate_count: int = 2):
	"""
	Template: Create VCF file for each 'chromosome' in pooled alignment.
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	bam_file_string = ' -b '.join(bam_file_list)
	inputs = {'reference': reference_genome_file,
		   	  'bam': bam_file_list}
	outputs = {'vcf': f'{output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.vcf'}
	options = {
		'cores': 1,
		'memory': '80g',
		'walltime': '60:00:00'
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	export _JAVA_OPTIONS="-Xmx{options['memory']}"

	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"

	[ -d {output_directory}/freebayes_vcf/tmp ] || mkdir -p {output_directory}/freebayes_vcf/tmp
	
	freebayes \
		-f {reference_genome_file} \
		-n {best_n_alleles} \
		-p {ploidy} \
		-r {region}:{start}-{end} \
		--min-alternate-fraction {min_alternate_fraction} \
		--min-alternate-count {min_alternate_count} \
		--pooled-discrete \
		-b {bam_file_string} \
		> {output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.prog.vcf
	
	mv {output_directory}/freebayes_vcf/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region}.prog.vcf {outputs['vcf']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def index_reference_genome(reference_genome_file: str, output_directory: str):
	"""
	Template: Index reference genome with :script:`bwa index` and :script:`samtools faidx`.
	
	Template I/O::
	
		inputs = {'reference': reference_genome_file}
		outputs = {'symlink': *, 'bwa': [*.amb, *.ann, *.pac, *.bwt, *.sa], 'fai': *.fai}
	
	:param str reference_genome_file:
		Path to reference genome.
	:param str output_directory:
		Path to output directory.
	"""
	inputs = {'reference': reference_genome_file}
	outputs = {'symlink': f'{output_directory}/reference/{os.path.basename(reference_genome_file)}',
			   'bwa': [f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.amb',
					   f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.ann',
					   f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.pac',
					   f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.bwt',
					   f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.sa'],
			   'fai': f'{output_directory}/reference/{os.path.basename(reference_genome_file)}.fai'}
	protect = [outputs['symlink'], outputs['bwa'][0], outputs['bwa'][1], outputs['bwa'][2], outputs['bwa'][3], outputs['bwa'][4], outputs['fai']]
	options = {
		'cores': 1,
		'memory': '10g',
		'walltime': '04:00:00'
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/reference ] || mkdir -p {output_directory}/reference
	[ -e {output_directory}/reference/{os.path.basename(reference_genome_file)} ] && rm -f {output_directory}/reference/{os.path.basename(reference_genome_file)}
	ln -s {reference_genome_file} {output_directory}/reference/{os.path.basename(reference_genome_file)}
	
	bwa index \
		-p {output_directory}/reference/{os.path.basename(reference_genome_file)} \
		{output_directory}/reference/{os.path.basename(reference_genome_file)}
	
	samtools faidx \
		-o {output_directory}/reference/{os.path.basename(reference_genome_file)}.prog.fai \
		{output_directory}/reference/{os.path.basename(reference_genome_file)}
	
	mv {output_directory}/reference/{os.path.basename(reference_genome_file)}.prog.fai {outputs['fai']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, protect=protect, options=options, spec=spec)

def depth_distribution(bam_files: list, output_directory: str, species_name: str):
	"""
	Template: Calculate the per site depth distribution across all populations using :script:`samtools depth`.
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'bam': bam_files}
	outputs = {'depth': f'{output_directory}/depth_distribution/{species_abbreviation(species_name)}.depth'}
	options = {
		'cores': 30,
		'memory': '20g',
		'walltime': '12:00:00'
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/depth_distribution ] || mkdir -p {output_directory}/depth_distribution
	
	samtools depth \
		--threads {options['cores']} \
		-o {output_directory}/depth_distribution/{species_abbreviation(species_name)}.prog.depth \
		{' '.join(bam_files)}
	
	mv {output_directory}/depth_distribution/{species_abbreviation(species_name)}.prog.depth {outputs['depth']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def depth_distribution_plot(depth_distribution_file: str, min_coverage_threshold: int, output_directory: str, plot_depth_distribution: str = f'{os.path.dirname(os.path.realpath(__file__))}/software/depthdistribution.py'):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'depth': depth_distribution_file}
	outputs = {'plot': f'{output_directory}/depth_distribution/{os.path.basename(depth_distribution_file)}.png',
			   'tsv': f'{output_directory}/depth_distribution/{os.path.basename(depth_distribution_file)}.tsv'}
	options = {
		'cores': 1,
		'memory': '100g',
		'walltime': '04:00:00'
	}
	protect = [outputs['plot'], outputs['tsv']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/depth_distribution ] || mkdir -p {output_directory}/depth_distribution

	python {plot_depth_distribution} \
		{min_coverage_threshold} \
		{depth_distribution_file} \
		{output_directory}/depth_distribution

	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def shared_sites_within_threshold_bed(depth_distribution_file: str, depth_distribution_tsv: str, output_directory: str, species_name: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'depth_file': depth_distribution_file,
		   	  'depth_tsv': depth_distribution_tsv}
	outputs = {'bed': f'{output_directory}/depth_distribution/{species_abbreviation(species_name)}.depththreshold.bed'}
	options = {
		'cores': 1,
		'memory': '10g',
		'walltime': '02:00:00'
	}
	protect = [outputs['bed']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/depth_distribution ] || mkdir -p {output_directory}/depth_distribution
	
	bedtools merge \
		-i <(awk \
		-v maxthreshold=$(awk 'BEGIN{{FS = OFS = "\\t"}} {{if (NR == 2) {{print $8; exit}}}}' {depth_distribution_tsv}) \
		-v minthreshold=$(awk 'BEGIN{{FS = OFS = "\\t"}} {{if (NR == 2) {{print $7; exit}}}}' {depth_distribution_tsv}) \
		'BEGIN{{
			FS = OFS = "\\t"
		}}
		{{
			checksum = 0
			for (i = 3; i <= NF; i++)
				{{
					if ($i >= minthreshold && $i <= maxthreshold)
						{{checksum += 1}}
				}}
			if (checksum == NF - 2)
				{{
					print $1, $2 - 1, $2
				}}
		}}' \
		{depth_distribution_file}) \
		> {output_directory}/depth_distribution/{species_abbreviation(species_name)}.depththreshold.prog.bed
	
	mv {output_directory}/depth_distribution/{species_abbreviation(species_name)}.depththreshold.prog.bed {outputs['bed']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def name_freebayes_partition_single(idx: str, target: AnonymousTarget) -> str:
	return f'freebayes_part_single_{os.path.basename(target.outputs["vcf"]).replace("-", "_").replace("|", "_")}'

def freebayes_partition_single(reference_genome_file: str, bam_file: str, output_directory: str, group_name: str, sample_name: str, region: str, num: int, start: int, end: int, ploidy: int = 100, best_n_alleles: int = 3, min_alternate_fraction: float | int = 0, min_alternate_count: int = 2, memory: int = 80, time: str = '48:00:00'):
	"""
	Template: Create VCF file for each partition in a single pooled alignment.
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'reference': reference_genome_file,
		   	  'bam': bam_file}
	outputs = {'vcf': f'{output_directory}/raw_vcf/{group_name}/{sample_name}/tmp/{sample_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz',
			   'index': f'{output_directory}/raw_vcf/{group_name}/{sample_name}/tmp/{sample_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz.csi'}
	options = {
		'cores': 1,
		'memory': f'{memory}g',
		'walltime': time
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	export _JAVA_OPTIONS="-Xmx{options['memory']}"

	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"

	[ -d {output_directory}/raw_vcf/{group_name}/{sample_name}/tmp ] || mkdir -p {output_directory}/raw_vcf/{group_name}/{sample_name}/tmp
	
	freebayes \
		--fasta-reference {reference_genome_file} \
		--use-best-n-alleles {best_n_alleles} \
		--ploidy {ploidy} \
		--region '{region}:{start}-{end}' \
		--min-alternate-fraction {min_alternate_fraction} \
		--min-alternate-count {min_alternate_count} \
		--pooled-discrete \
		-b {bam_file} \
	| bcftools view \
		--output-type z \
		--output {output_directory}/raw_vcf/{group_name}/{sample_name}/tmp/{sample_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz \
		--write-index \
		-
	
	mv {output_directory}/raw_vcf/{group_name}/{sample_name}/tmp/{sample_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/raw_vcf/{group_name}/{sample_name}/tmp/{sample_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz.csi {outputs['index']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def name_freebayes_partition_group(idx: str, target: AnonymousTarget) -> str:
	return f'freebayes_part_group_{os.path.basename(target.outputs["vcf"]).replace("-", "_").replace("|", "_")}'

def freebayes_partition_group(reference_genome_file: str, bam_files: list, output_directory: str, species_name: str, group_name: str, region: str, num: int, start: int, end: int, ploidy: int = 100, best_n_alleles: int = 3, min_alternate_fraction: float | int = 0, min_alternate_count: int = 2, memory: int = 80, time: str = '48:00:00'):
	"""
	Template: Create VCF file for each partition in a group of pooled alignments.
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'reference': reference_genome_file,
		   	  'bam': bam_files}
	outputs = {'vcf': f'{output_directory}/raw_vcf/{group_name}/tmp/{species_abbreviation(species_name)}_{group_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz',
			   'index': f'{output_directory}/raw_vcf/{group_name}/tmp/{species_abbreviation(species_name)}_{group_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz.csi'}
	options = {
		'cores': 1,
		'memory': f'{memory}g',
		'walltime': time
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	export _JAVA_OPTIONS="-Xmx{options['memory']}"

	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"

	[ -d {output_directory}/raw_vcf/{group_name}/tmp ] || mkdir -p {output_directory}/raw_vcf/{group_name}/tmp
	
	freebayes \
		--fasta-reference {reference_genome_file} \
		--use-best-n-alleles {best_n_alleles} \
		--ploidy {ploidy} \
		--region '{region}:{start}-{end}' \
		--min-alternate-fraction {min_alternate_fraction} \
		--min-alternate-count {min_alternate_count} \
		--pooled-discrete \
		-b {' -b '.join(bam_files)} \
	| bcftools view \
		--output-type z \
		--output {output_directory}/raw_vcf/{group_name}/tmp/{species_abbreviation(species_name)}_{group_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz \
		--write-index \
		-
	
	mv {output_directory}/raw_vcf/{group_name}/tmp/{species_abbreviation(species_name)}_{group_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/raw_vcf/{group_name}/tmp/{species_abbreviation(species_name)}_{group_name}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz.csi {outputs['index']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def name_freebayes_partition_all(idx: str, target: AnonymousTarget) -> str:
	return f'freebayes_part_all_{os.path.basename(target.outputs["vcf"]).replace("-", "_").replace("|", "_")}'

def freebayes_partition_all(reference_genome_file: str, bam_files: list, output_directory: str, species_name: str, region: str, num: int, start: int, end: int, ploidy: int = 100, best_n_alleles: int = 3, min_alternate_fraction: float | int = 0, min_alternate_count: int = 2, memory: int = 80, time: str = '48:00:00'):
	"""
	Template: Create VCF file for each partition in a large set of pooled alignments.
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'reference': reference_genome_file,
		   	  'bam': bam_files}
	outputs = {'vcf': f'{output_directory}/raw_vcf/all/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz',
			   'index': f'{output_directory}/raw_vcf/all/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.vcf.gz.csi'}
	options = {
		'cores': 1,
		'memory': f'{memory}g',
		'walltime': time
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	export _JAVA_OPTIONS="-Xmx{options['memory']}"

	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"

	[ -d {output_directory}/raw_vcf/all/tmp ] || mkdir -p {output_directory}/raw_vcf/all/tmp
	
	freebayes \
		--fasta-reference {reference_genome_file} \
		--use-best-n-alleles {best_n_alleles} \
		--ploidy {ploidy} \
		--region '{region}:{start}-{end}' \
		--min-alternate-fraction {min_alternate_fraction} \
		--min-alternate-count {min_alternate_count} \
		--pooled-discrete \
		-b {' -b '.join(bam_files)} \
	| bcftools view \
		--output-type z \
		--output {output_directory}/raw_vcf/all/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz \
		--write-index \
		-
	
	mv {output_directory}/raw_vcf/all/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/raw_vcf/all/tmp/{species_abbreviation(species_name)}.freebayes_n{best_n_alleles}_p{ploidy}_minaltfrc{min_alternate_fraction}_minaltcnt{min_alternate_count}.{num}_{region.replace("|", "_")}.prog.vcf.gz.csi {outputs['index']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def concat(files: list, output_name: str, output_directory: str | None = None, compress: bool = False):
	"""
	Template: Name-sorts and concatenates files. Optionally compresses output using :script:`gzip`.
	
	Template I/O::
	
		inputs = {'files': files}
		outputs = {'concat_file': output_name.ext | output_name.ext.gzip}
	
	:param list files:
		List containing files to concatenate.
	:param str output_name:
		Desired name of output file, no extension.
	:param str output_directory:
		Path to output directory. Default is directory of 'files[0]'.
	:param bool compress:
		Bool indicating whether the output file should be compressed or not.
	"""
	if output_directory is None:
		output_directory = os.path.dirname(files[0])
	inputs = {'files': files}
	if compress:
		outputs = {'concat_file': f'{output_directory}/{output_name}{os.path.splitext(files[0])[1]}.gz'}
	else:
		outputs = {'concat_file': f'{output_directory}/{output_name}{os.path.splitext(files[0])[1]}'}
	options = {
		'cores': 2,
		'memory': '16g',
		'walltime': '24:00:00'
	}
	protect = outputs['concat_file']
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}] || mkdir -p {output_directory}

	if [ {compress} == 'False' ]; then
		cat \
			{' '.join(files)} \
			> {output_directory}/{output_name}.prog{os.path.splitext(files[0])[1]}
		
		mv {output_directory}/{output_name}.prog{os.path.splitext(files[0])[1]} {outputs['concat_file']}
	else
		cat \
			{' '.join(files)} \
		| gzip \
			-c \
			- \
			> {output_directory}/{output_name}.prog{os.path.splitext(files[0])[1]}.gz
		
		mv {output_directory}/{output_name}.prog{os.path.splitext(files[0])[1]}.gz {outputs['concat_file']}
	fi

	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def concat_vcf(files: list, output_name: str, output_directory: str| None = None, compress: bool = True):
	"""
	Template: Concatenates :format:`VCF` files. Optionally compresses output.
	
	Template I/O::
	
		inputs = {'files': files}
		outputs = {'concat_file': output_name.ext | output_name.ext.gzip}
	
	:param list files:
		List containing :format:`VCF` files to concatenate. Can already be gzipped.
	:param str output_name:
		Desired name of output file, no extension.
	:param str output_directory:
		Path to output directory. Default is directory of 'files[0]'.
	:param bool compress:
		Bool indicating whether the output file should be compressed or not.
	"""
	if not output_directory:
		output_directory = os.path.dirname(files[0])
	inputs = {'files': files}
	if compress:
		outputs = {'concat_file': f'{output_directory}/{output_name}.vcf.gz',
			 	   'index': f'{output_directory}/{output_name}.vcf.gz.csi'}
	else:
		outputs = {'concat_file': f'{output_directory}/{output_name}.vcf'}
	options = {
		'cores': 32,
		'memory': '40g',
		'walltime': '24:00:00'
	}
	protect = outputs['concat_file']
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory} ] || mkdir -p {output_directory}
	
	if [ {compress} == 'False' ]; then
		bcftools concat \
			--threads {options['cores']} \
			--output-type v \
			--output {output_directory}/{output_name}.prog.vcf \
			{' '.join(files)}

			mv {output_directory}/{output_name}.prog.vcf {outputs['concat_file']}
	else
		bcftools concat \
			--threads {options['cores']} \
			--output-type z \
			--output {output_directory}/{output_name}.prog.vcf.gz \
			{' '.join(files)}
		
		bcftools index \
			--threads {options['cores']} \
			{output_directory}/{output_name}.prog.vcf.gz

			mv {output_directory}/{output_name}.prog.vcf.gz {outputs['concat_file']}
			mv {output_directory}/{output_name}.prog.vcf.gz.csi {outputs['index']}
	fi
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def merge_and_norm_vcf(vcf_files: list, reference_genome_file: str, output_name: str, output_directory: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'vcfs': vcf_files}
	outputs = {'vcf': f'{output_directory}/{output_name}.merged.norm.vcf.gz',
			   'index': f'{output_directory}/{output_name}.merged.norm.vcf.gz.csi'}
	options = {
		'cores': 30,
		'memory': '40g',
		'walltime': '12:00:00'
	}
	protect = [outputs['vcf'], outputs['index']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory} ] || mkdir -p {output_directory}
	
	bcftools merge \
		--threads {options['cores']} \
		--output-type v \
		--missing-to-ref \
		{' '.join(vcf_files)} \
	| bcftools norm \
		--threads {options['cores']} \
		--output-type z \
		--output {output_directory}/{output_name}.merged.norm.prog.vcf.gz \
		--fasta-ref {reference_genome_file} \
		--write-index \
		-
	
	mv {output_directory}/{output_name}.merged.norm.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/{output_name}.merged.norm.prog.vcf.gz.csi {outputs['index']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def norm_vcf(vcf_file: str, reference_genome_file: str, output_name: str, output_directory: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'vcf': vcf_file}
	outputs = {'vcf': f'{output_directory}/{output_name}.norm.vcf.gz',
			   'index': f'{output_directory}/{output_name}.norm.vcf.gz.csi'}
	options = {
		'cores': 30,
		'memory': '40g',
		'walltime': '12:00:00'
	}
	protect = [outputs['vcf'], outputs['index']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory} ] || mkdir -p {output_directory}
	
	bcftools norm \
		--threads {options['cores']} \
		--output-type z \
		--output {output_directory}/{output_name}.norm.prog.vcf.gz \
		--fasta-ref {reference_genome_file} \
		--write-index \
		{vcf_file}
	
	mv {output_directory}/{output_name}.norm.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/{output_name}.norm.prog.vcf.gz.csi {outputs['index']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def site_count_region(bam_files: list, depth_distribution_tsv: str, site_type: str, output_directory: str, species_name: str, bed_file: str | None = None):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	if bed_file:
		inputs = {'bam': bam_files,
			   	  'depth': depth_distribution_tsv,
				  'bed': bed_file}
	else:
		inputs = {'bam': bam_files,
			   	  'depth': depth_distribution_tsv}
	outputs = {'sitetable': f'{output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.{site_type}.tsv'}
	options = {
		'cores': 10,
		'memory': '40g',
		'walltime': '06:00:00'
	}
	protect = [outputs['sitetable']]
	populations = '\034'.join([os.path.basename(i).split(sep=".")[0] for i in bam_files])
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/sitetable ] || mkdir -p {output_directory}/sitetable
	
	awk \
		-v maxthreshold=$(awk 'BEGIN{{FS = OFS = "\\t"}} {{if (NR == 2) {{print $8; exit}}}}' {depth_distribution_tsv}) \
		-v minthreshold=$(awk 'BEGIN{{FS = OFS = "\\t"}} {{if (NR == 2) {{print $7; exit}}}}' {depth_distribution_tsv}) \
		'BEGIN{{
			FS = OFS = "\\t"
		}}
		{{
			if (FNR == NR)
			{{
				n = split($0, populations, "\\034")
				nextfile
			}}
			checksum = 0
			for (i = 3; i <= NF; i++)
				{{
					if ($i >= minthreshold && $i <= maxthreshold)
						{{
							checksum += 1
							thressites[populations[i - 2], "whole_genome"] += 1
							thressites[populations[i - 2], $1] += 1
					}}
				}}
			if (checksum == NF - 2)
				{{
					thressites["all", "whole_genome"] += 1
					thressites["all", $1] += 1
				}}
			allsites["all", "whole_genome"] += 1
			allsites["all", $1] += 1
		}}
		END{{
			print "stage", "identifier", "group", "region", "type", "site_count"
			for (i in allsites)
			{{
				split(i, allsitesarray, "\\034")
				print "0", "total", allsitesarray[1], allsitesarray[2], "{site_type}", allsites[i]
			}}
			for (i in thressites)
			{{
				split(i, thressitesarray, "\\034")
				print "1", "within_threshold", thressitesarray[1], thressitesarray[2], "{site_type}", thressites[i]
			}}
		}}' \
		<(echo -e "{populations}") \
		<(samtools depth \
			--threads {options['cores']} \
			{f'-b {bed_file}' if bed_file else ''} \
			{' '.join(bam_files)}) \
	| awk \
		'{{
			if (NR == 1)
			{{
				print $0
				next
			}}
			print $0 | "sort -k 1,1 -k 3,3 -k 4,4"
		}}' \
		> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.{site_type}.prog.tsv

	mv {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.{site_type}.prog.tsv {outputs['sitetable']}

	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def extract_softmasked_intervals(reference_genome_file: str, output_directory: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'reference': reference_genome_file}
	outputs = {'bed': f'{output_directory}/annotation/{os.path.splitext(os.path.splitext(os.path.basename(reference_genome_file))[0])[0] if reference_genome_file.endswith('.gz') else os.path.splitext(os.path.basename(reference_genome_file))[0]}.repeats.bed'}
	options = {
		'cores': 1,
		'memory': '10g',
		'walltime': '02:00:00'
	}
	protect = {outputs['bed']}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/annotation ] || mkdir -p {output_directory}/annotation
	
	bedtools merge \
		-i <(awk \
			'function maskedinterval(string)
			{{
				result = match(string, /[a-z]+/, matcharray)
				if (result != 0)
				{{
					start = pos + RSTART - 1
					end = start + RLENGTH
					print seqname, start, end
					sub(matcharray[0], toupper(matcharray[0]), string)
 					maskedinterval(string)
				}}
			}}
			BEGIN{{
				FS = "\\t"
				OFS = "\\t"
			}}
			{{
				if ($0 ~ /^>/)
				{{
					split($1, seqnamearray, " ")
					seqname = substr(seqnamearray[1], 2)
					pos = 0
					next
				}}
				maskedinterval($0)
				pos += length($0)
			}}' \
			{'<(zcat' + reference_genome_file + ')' if reference_genome_file.endswith('.gz') else reference_genome_file}) \
		> {output_directory}/annotation/{os.path.splitext(os.path.splitext(os.path.basename(reference_genome_file))[0])[0] if reference_genome_file.endswith('.gz') else os.path.splitext(os.path.basename(reference_genome_file))[0]}.repeats.prog.bed
	
	mv {output_directory}/annotation/{os.path.splitext(os.path.splitext(os.path.basename(reference_genome_file))[0])[0] if reference_genome_file.endswith('.gz') else os.path.splitext(os.path.basename(reference_genome_file))[0]}.repeats.prog.bed {outputs['bed']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def bed_exclude_overlap(main_bed_file: str, subtraction_bed_file: str, output_directory: str, species_name: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'main': main_bed_file,
		   	  'sub': subtraction_bed_file}
	outputs = {'bed': f'{output_directory}/sitetable/{species_abbreviation(species_name)}.intergenic_excl_repeats.bed'}
	options = {
		'cores': 1,
		'memory': '10g',
		'walltime': '01:00:00'
	}
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/sitetable ] || mkdir -p {output_directory}/sitetable
	
	bedtools subtract \
		-a {main_bed_file} \
		-b {subtraction_bed_file} \
		> {output_directory}/sitetable/{species_abbreviation(species_name)}.intergenic_excl_repeats.prog.bed
	
	mv {output_directory}/sitetable/{species_abbreviation(species_name)}.intergenic_excl_repeats.prog.bed {outputs['bed']}
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, spec=spec)

def filter_vcf(vcf_file: str, depth_distribution_tsv: str, output_directory: str, species_name: str, min_depth: int = 200):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'vcf': vcf_file,
		   	  'depth': depth_distribution_tsv}
	outputs = {'vcf': f'{output_directory}/{os.path.splitext(os.path.splitext(os.path.basename(vcf_file))[0])[0] if vcf_file.endswith(".gz") else os.path.splitext(os.path.basename(vcf_file))[0]}.bcftoolsfilter_AF0_SnpGap5_typesnps_biallelic_DP{min_depth}-dynamic_AO1.vcf.gz',
			   'index': f'{output_directory}/{os.path.splitext(os.path.splitext(os.path.basename(vcf_file))[0])[0] if vcf_file.endswith(".gz") else os.path.splitext(os.path.basename(vcf_file))[0]}.bcftoolsfilter_AF0_SnpGap5_typesnps_biallelic_DP{min_depth}-dynamic_AO1.vcf.gz.csi',
			   'sitetable': f'{output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.tsv'}
	options = {
		'cores': 18,
		'memory': '30g',
		'walltime': '12:00:00'
	}
	protect = [outputs['vcf'], outputs['index'], outputs['sitetable']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/sitetable ] || mkdir -p {output_directory}/sitetable

	variablesitecount() {{
		awk \
			-v first="$1" \
			-v stage="$2" \
			-v identifier="$3" \
			'BEGIN{{
				FS = OFS = "\\t"
			}}
			{{
				if ($0 ~ /^##/)
				{{
					next
				}}
				if ($0 ~ /^#CHROM/)
				{{
					for (i = 10; i <= NF; i++)
					{{
						populations[i] = $i
					}}
					next
				}}
				for (i = 10; i <= NF; i++)
				{{
					split($i, genotypearray, ":")
					if (genotypearray[1] !~ /^0\\/.*0$/)
					{{
						npopulationvariants[populations[i], "whole_genome"] += 1
						npopulationvariants[populations[i], $1] += 1
					}}
				}}
				ntotalvariants["whole_genome"] += 1
				ntotalvariants[$1] += 1
			}}
			END{{
				if (first == 1)
				{{
					print "stage", "identifier", "group", "region", "type", "site_count"
				}}
				for (i in ntotalvariants)
				{{
					print stage, identifier, "all", i, "variable", ntotalvariants[i]
				}}
				for (i in npopulationvariants)
				{{
					split(i, popregion, "\\034")
					print stage, identifier, popregion[1], popregion[2], "variable", npopulationvariants[i]
				}}
			}}'
	}}
	
	maxdepth=$(awk \
		'BEGIN{{
			FS = OFS = "\\t"
		}}
		{{
			if (NR == 2)
			{{
				print $8
				exit
			}}
		}}' \
		{depth_distribution_tsv})
	
	bcftools view \
		--threads {options['cores']} \
		--output-type v \
		{vcf_file} \
	| tee \
		>(variablesitecount \
			1 \
			0 \
			"total" \
			> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--include 'INFO/AF > 0' \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			2 \
			"AF>0" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools filter \
		--threads {options['cores']} \
		--SnpGap 5:indel \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			3 \
			"indel_proximity" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--types snps \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			4 \
			"snps_only" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--max-alleles 2 \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			5 \
			"biallelic_only" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--include "FMT/DP>={min_depth} & FMT/DP<=$maxdepth" \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			6 \
			"depth_thresholds" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--include 'AVG(FMT/AO) > 1' \
		--output-type v \
		- \
	| tee \
		>(variablesitecount \
			0 \
			7 \
			"AO>1" \
			>> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv) \
	| bcftools view \
		--threads {options['cores']} \
		--output-type z \
		--output {output_directory}/{os.path.splitext(os.path.splitext(os.path.basename(vcf_file))[0])[0] if vcf_file.endswith(".gz") else os.path.splitext(os.path.basename(vcf_file))[0]}.bcftoolsfilter_AF0_SnpGap5_typesnps_biallelic_DP{min_depth}-dynamic_AO1.prog.vcf.gz \
		--write-index \
		-
	
	awk \
		'BEGIN{{
			FS = OFS = "\\t"
		}}
		{{
			if (NR == 1)
			{{
				print $0
				next
			}}
			print $0 | "sort -k 1,1 -k 3,3 -k 4,4"
		}}' \
		{output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv \
		> {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.prog.tsv
		
	mv {output_directory}/{os.path.splitext(os.path.splitext(os.path.basename(vcf_file))[0])[0] if vcf_file.endswith(".gz") else os.path.splitext(os.path.basename(vcf_file))[0]}.bcftoolsfilter_AF0_SnpGap5_typesnps_biallelic_DP{min_depth}-dynamic_AO1.prog.vcf.gz {outputs['vcf']}
	mv {output_directory}/{os.path.splitext(os.path.splitext(os.path.basename(vcf_file))[0])[0] if vcf_file.endswith(".gz") else os.path.splitext(os.path.basename(vcf_file))[0]}.bcftoolsfilter_AF0_SnpGap5_typesnps_biallelic_DP{min_depth}-dynamic_AO1.prog.vcf.gz.csi {outputs['index']}
	mv {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.prog.tsv {outputs['sitetable']}
	rm {output_directory}/sitetable/{species_abbreviation(species_name)}.sitetable.variable.unsorted.tsv
	
	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)

def merge_site_tables(site_tables: list, output_name: str, output_directory: str):
	"""
	Template: template_description
	
	Template I/O::
	
		inputs = {}
		outputs = {}
	
	:param
	"""
	inputs = {'tables': site_tables}
	outputs = {'sitetable': f'{output_directory}/sitetable/{output_name}.sitetable.tsv'}
	options = {
		'cores': 1,
		'memory': '10g',
		'walltime': '02:00:00'
	}
	protect = [outputs['sitetable']]
	spec = f"""
	# Sources environment
	if [ "$USER" == "jepe" ]; then
		source /home/"$USER"/.bashrc
		source activate popgen
	fi
	
	echo "START: $(date)"
	echo "JobID: $SLURM_JOBID"
	
	[ -d {output_directory}/sitetable ] || mkdir -p {output_directory}/sitetable
	
	awk \
		'BEGIN{{
			FS = OFS = "\\t"
		}}
		{{
			if (FNR == 1)
			{{
				if (FNR == NR)
				{{
					print $0
				}}
				next
			}}
			print $0 | "sort -k 1,1 -k 3,3 -k 4,4"
		}}' \
		{' '.join(site_tables)} \
		> {output_directory}/sitetable/{output_name}.sitetable.prog.tsv
	
	mv {output_directory}/sitetable/{output_name}.sitetable.prog.tsv {outputs['sitetable']}

	echo "END: $(date)"
	echo "$(jobinfo "$SLURM_JOBID")"
	"""
	return AnonymousTarget(inputs=inputs, outputs=outputs, options=options, protect=protect, spec=spec)